drivers: i2s: Add support for I2S stm32

avisconti opened the pull request:
STM32 I2S driver implementation. It has been designed in the most
generic way possible, with the possibility of using it in master/slave
and rx/tx mode. Currenty it has been tested for master rx mode only
using the microphone on ArgonKey board.

The configuration file permits to compile it for STM32F4xx product family
only, but it should be easy to extend it also for other families.

It supports all 5 STM32F4xx I2S controllers (I2S 1/4/5 on APB2 and I2S 2/3
on APB1). It makes uses of the available DMA channels for rx/tx streams.

The clock source can be selected among one of the following two choices:

  - PLLI2S pll, with possibility to configure PLLM/PLLN/PLLR
  - HSE/HSI clock

Interrupt is triggered only in case of errors (FRM/OVR/UDR).

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
drivers: i2s: Add support for I2S stm32

STM32 I2S driver implementation. It has been designed in the most
generic way possible, with the possibility of using it in master/slave
and rx/tx mode. Currenty it has been tested for master rx mode only
using the microphone on ArgonKey board.

The configuration file permits to compile it for STM32F4xx product
family only, but it should be easy to extend it also for other
families.

It supports all 5 STM32F4xx I2S controllers (I2S 1/4/5 on APB2 and
I2S 2/3 on APB1).
It makes uses of the available DMA channels for rx/tx streams.

The clock source can be selected among one of the following two choices:

  - PLLI2S pll, with possibility to configure PLLM/PLLN/PLLR
  - HSE/HSI clock

Interrupt is triggered only in case of errors (FRM/OVR/UDR).

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@erwango 
Would be nice if you could take a look at it.

avisconti commented:
recheck

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1170 @@
+		.bus = STM32_CLOCK_BUS_APB2,
+	},
+#ifdef CONFIG_I2S_STM32_USE_PLLI2S_ENABLE
+	.i2s_clk_sel = LL_RCC_I2S2_CLKSOURCE_PLLI2S,
```
LL_RCC_I2S1_CLKSOURCE_PLLI2S?

avisconti replied:
No, this one is correct. But thanks to your observation I was able to find a couple of mistakes in others. So, basically:
 SPI/I2S 1/4/5 are on APB2 and must use LL_RCC_I2S2_CLKSOURCE_PLLI2S
 SPI/I2S 2/3 are on APB1 and must use LL_RCC_I2S1_CLKSOURCE_PLLI2S

I noticed that in my code SPI/I2S 2/3 are wrong. So, thanks!

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1170 @@
+static void i2s_stm32_irq_config_func_1(struct device *dev);
+
+static const struct i2s_stm32_cfg i2s_stm32_config_1 = {
+	.i2s = (SPI_TypeDef *) CONFIG_SPI_1_BASE_ADDRESS,
```
How does this work with device tree?
It means that SPI_1 should be enabled ?

erwango replied:
Well, in that case, why not doing it in this PR? It should not be a major effort.
Besides having several IPs located at same address make it an interesting test for dts scripts.

avisconti replied:
Well, this was just because I'm currently not very familiar with dts stuff in zephyr. Nevertheless,I agree it should not be so huge effort. I probably would need some help from your side. For the moment I'm just pushing a new commit with the other changes. I'll come back on this dts topic soon...

avisconti replied:
@mnkp 
I think that @erwango is correct.
Yesterday I was compiling this driver for Nucleo, with I2S enabled but SPI disabled.
And it does not work because it does not find CONFIG_SPI_1_BASE_ADDRESS.
At the moment it is needed to enable both I2S and SPI to make it working, while I guess
it is not expected. As a user I would like to enable I2S only.

mnkp replied:
The message I was trying to convey is a little bit different. The name of the hardware module according to the documentation is SPI and we probably need only one representation of this module in DTS. There is no really need to define I2S in DTS. At least for now. From the hardware point of view it is the same peripheral, same base address, IRQs, etc. 

And yes, to use I2S driver we would need to ensure that necessary instance of SPI hardware module is enabled in DTS otherwise all the defines generated by DTS will be missing.

Later when we refine DTS concept and have automatic generation of driver instances in driver source code this will need to change. But from what I've noticed no one knows yet how to do it properly, the discussion is ongoing, so it's a little bit too early to try to implement it.

erwango replied:
@mnkp 

> There is no really need to define I2S in DTS [...] it's a little bit too early to try to implement it.

There is already some part of DTS used in a lot of STM32 drivers, such as UART, SPI, I2C, ...
This is not generated instantiation yet, but at least we are able to gather configuration bits
from dts files and input to drivers with the help of .fixup files.
Request for this driver is to achieve the same status.

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1170 @@
+#ifdef CONFIG_I2S_STM32_USE_PLLI2S_ENABLE
+	.i2s_clk_sel = LL_RCC_I2S2_CLKSOURCE_PLLI2S,
+#else
+	.i2s_clk_sel = LL_RCC_I2S2_CLKSOURCE_PLLSRC,
```
LL_RCC_I2S1_CLKSOURCE_PLLSRC?

avisconti replied:
see above

erwango commented on drivers/i2s/Kconfig.stm32:
```diff
@@ -0,0 +1,67 @@
+	select DMA
+	default n
+	help
+	  Enable I2S support on the STM32 family of processors.
```
Please indicate here on which series it was tested

avisconti replied:
ack

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1170 @@
+struct queue_item tx_1_ring_buf[CONFIG_I2S_STM32_TX_BLOCK_COUNT + 1];
+
+static struct i2s_stm32_data i2s_stm32_data_1 = {
+	.dma_name = I2S1_DMA_NAME,
```
Would be great to get most of this from DMA.
Though, IIUC, this mostly requires DMA node under dt, is that correct?

avisconti replied:
That's correct!
I candidate myself to add DMA in DT in the future if you do not have time. But actually I would also need to understand bit more of the zephyr DT structure.

mnkp reviewed (changes requested):
I have a question, did you try to run the testcases in tests/drivers/i2s on your target board? If stm32 devices do not support internal loopback you may need to create an external one. In any case, the tests would ensure that there are no major problems with this driver.

avisconti commented:
@mnkp 
> I have a question, did you try to run the testcases in tests/drivers/i2s on your target board? If stm32 
> devices do not support internal loopback you may need to create an external one. In any case, the 
> tests would ensure that there are no major problems with this driver.

It will be very difficult on my setup. The ArgonKey is not designed to do debug and it is
just limited in the pinout.
One possibility would be to use a different board. @erwango, what do you think?
Do you have a suitable board I can borrow? Which board would be the best for such a test?

avisconti commented on drivers/i2s/Kconfig:
```diff
@@ -35,6 +35,76 @@ config SYS_LOG_I2S_LEVEL
 	  3 INFO, write SYS_LOG_INF in addition to previous levels
 	  4 DEBUG, write SYS_LOG_DBG in addition to previous levels
 
+config	I2S_1
```
The reason why I did it was because I saw it done in this way in SPI driver. I just prepared the new commit for the PR. I will test it out just a bit before pushing out.

mnkp commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1171 @@
+	return 0;
+}
+
+static inline int i2s_stm32_enable_clock(struct device *dev)
```
We generally use `static inline` only if we want to place the function in the header file. In this case the `inline` is not required. It's better to let the compiler decide on how to run optimizations. Without `inline` the code is easier to maintain. This function is currently called only once, however it may change in the future.

Using `static inline` in .c file for short functions is probably still OK, it documents the intent for users reading the code.

avisconti replied:
OK, it makes sense.

mnkp commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1171 @@
+
+#endif /* CONFIG_I2S_5 */
+
+static inline struct device *get_dev_from_rx_dma_channel(u32_t dma_channel)
```
That's a bit inconsistent. We generally don't place any functions after the device instantiation code. Let's move it somewhere above 
```
#ifdef CONFIG_I2S_1
```
line

avisconti replied:
ack

mnkp commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1171 @@
+
+static int rx_stream_start(struct stream *stream, struct device *dev)
+{
+	const struct i2s_stm32_cfg *cfg = dev->config->config_info;
```
Since we use DEV_DATA macro, shouldn't we use here the DEV_CFG one? Like in `i2s_stm32_initialize` function. Several other functions are also affected.

avisconti replied:
Oooops.... correct!
Thanks for catching it out.

mnkp replied:
Thanks for fixing, however there are still a couple of functions which don't use DEV_CFG macro.

avisconti replied:
sigh... I'm doing too many stuff at the same time.
Thanks.

mnkp reviewed (commented):
Thanks for the fixes. I have a few more minor comments. The one from @erwango are more important. Otherwise the code looks good.

mnkp commented:
The code LGTM, however - as I already mentioned - we really should stress test this driver with the I2S testcase. It's a pretty complex code.

avisconti committed:
arch: arm: soc: stm32f4xx: Add LL include file also for I2S case

stm32f4xx_ll_spi.h include file is shared between I2S and SPI
(controller is same). So we need to include it even when I2S_STM32
is enabled.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
As requested by @erwango I added another small commit to this PR.
This commit is fixing an issue in stm32f4xx  soc.h, where the stm32f4xx_ll_spi.h file is included only if SPI_STM32 is enabled, while it should be also if I2S_STM32 is enabled.

mnkp commented:
Github has problems with displaying commits in the proper order and maybe everything is fine. But it looks like the first commit "drivers: i2s: Add support for I2S stm32" will not compile unless the second one is present. So we should either reverse the order of the two commits or probably event better merge them.

avisconti commented:
@mnkp 

> Github has problems with displaying commits in the proper order and maybe 
> everything is fine. But it looks like the first commit 
> "drivers: i2s: Add support for I2S stm32" will not compile unless the second one is present? 
> So we should either reverse the order of the two commits or probably event better merge them.

Well, actually I made sure in i2s-stm32-develop branch that the soc.h fix was before the i2s driver. And I did it exactly for the reason you mentioned, i.e. not breaking compilation.

So, I'm not sure about github commit order, but it seems to me that it is showed in the same way 'git log' does. Hence, it should be correct. Does it make sens to you?

avisconti commented:
bash> git log --oneline i2s-stm32-develop
425c062 drivers: i2s: Add support for I2S stm32
b2ffa5f arch: arm: soc: stm32f4xx: Add LL include file also for I2S case

mnkp commented:
@avisconti yes, that looks good.

avisconti commented:
@erwango @mnkp 
I found another issue while compiling for nucleo_f411 and I realize that the F4XX family does not share the same clocktree among the various SoC variants. I read the stm32f4xx_ll_rcc.h include file and I was able to see that there are 3 cases:

1. if defined(RCC_CFGR_I2SSRC)
(stm32f469xx stm32f439xx stm32f427xx stm32f405xx stm32f401xe stm32f437xx stm32f429xx stm32f415xx stm32f407xx stm32f411xe stm32f417xx stm32f479xx stm32f401xc)

    In this case there is a single selector for the I2S clock source:
        -  LL_RCC_I2S1_CLKSOURCE_PLLI2S
        -  LL_RCC_I2S1_CLKSOURCE_PIN

2. if defined(RCC_DCKCFGR_I2SSRC)
(stm32f410tx stm32f410rx stm32f410cx)

    In this case there is a single selector for the I2S clock source:
        -  LL_RCC_I2S1_CLKSOURCE_PLL
        -  LL_RCC_I2S1_CLKSOURCE_PIN
        -  LL_RCC_I2S1_CLKSOURCE_PLLSRC (this is HSI/HSE)

3. if defined(RCC_DCKCFGR_I2S1SRC) and defined(RCC_DCKCFGR_I2S2SRC)
(stm32f423xx stm32f413xx stm32f446xx stm32f412cx stm32f412rx stm32f412zx stm32f412vx)

    In this case there are two selectors for the I2S clock source.
        sel_1.
        -  LL_RCC_I2S1_CLKSOURCE_PLLI2S
        -  LL_RCC_I2S1_CLKSOURCE_PIN
        -  LL_RCC_I2S1_CLKSOURCE_PLL
        -  LL_RCC_I2S1_CLKSOURCE_PLLSRC (this is HSI/HSE)
        sel_2.
        -  LL_RCC_I2S2_CLKSOURCE_PLLI2S
        -  LL_RCC_I2S2_CLKSOURCE_PIN
        -  LL_RCC_I2S2_CLKSOURCE_PLL
        -  LL_RCC_I2S2_CLKSOURCE_PLLSRC (this is HSI/HSE)

So, based on for which STM32F4XX platform you compile the driver needs to use the proper definition. I identified two possible ways, for which I may take your inputs (provided that you arrived to read up to here):

Solution 1: (PREFERRED) use the 3 RCC definitions to discriminate
Solution2: use the SOC names (too many) to discriminate

mnkp commented:
> Solution 1: (PREFERRED) use the 3 RCC definitions to discriminate
> Solution2: use the SOC names (too many) to discriminate

For me testing defines sourced from the soc header files (Solution 1) is also the most reliable way to handle this. Though probably one of the maintainers @galak, @MaureenHelm should have the last word.

erwango commented:
@avisconti 

Whatever being the chosen solution, even if F4 is the only impacted series for now, please have a check on how this scales on other series.
Solution 1 sounds like the easiest and maintainable for now. Though, I'd like to see how it fits with dts.

In the end this will be set from dts, so ideally solution should be close from what could be done on a fully deployed dts solution. I see several options with dts:
- Get a unified clock setting from dts, then combined with CMSIS files definitions in the driver to use the correct define
- Set precisely the clock setting in dts files (in this case we can adapt precisely, per SoC, according to each SoC IP). Then driver should be compatible with the various possible settings (one compatible solution)
- Having several compatibles and  depending on the compatible defined in dts have a dedicated driver implementation (more complex to implement today, but some settings are already handled this way in STM32 I2C and SPI)

avisconti commented:
@erwango 
I'm back to this topic after vacation + other urgent stuff

> Whatever being the chosen solution, even if F4 is the only impacted series for now, 
> please have a check on how this scales on other series.
> Solution 1 sounds like the easiest and maintainable for now.

Yes, I'm going with sol1. Please note that for other series would be similar, but not exactly the
same. I do not think that there will be an easy way w/o using ifdefs with the family name.

> Though, I'd like to see how it fits with dts.
> In the end this will be set from dts, so ideally solution should be close from what could be 
> done on a fully deployed dts solution. 

I guess it will be totally different at this point.
I would say that defining the clock in the dts will be a huge work if we want to have a work independent from the architecture.

I will push a new commit set asap.

avisconti committed:
soc/arm/stm32f4xx: Add I2S defines to dts.fixup

For the stm32f4xx family the SPI and I2S are sharing same
controller, so we can reuse same definitions.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
dts/arm/st/stm32f4: Add I2S support to DTS

Add I2S bus initial support to stm32f4 SOC family.
On stm32f4 the I2S shares same controller as SPI, so
the declarations have been adapted from SPI ones.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@erwango 
I just submitted a new version addressing following two points:
 1. I changed clocktree using Solution 1 (see above)
 2. I added I2S to stm32f4 family DTS

avisconti commented:
rebased to latest commit (there were some conflicts on DTS)

erwango commented on soc/arm/st_stm32/stm32f4/dts.fixup:
```diff
@@ -80,6 +80,36 @@
 #define CONFIG_SPI_6_NAME			ST_STM32_SPI_40015400_LABEL
 #define CONFIG_SPI_6_IRQ			ST_STM32_SPI_40015400_IRQ_0
 
+#define CONFIG_I2S_1_BASE_ADDRESS               ST_STM32_I2S_40013000_BASE_ADDRESS
```
Nitpick: Can you align all these ?
Seems #1 and #2 are one or 2 chars away from others

avisconti replied:
This file is using sometimes tabs and sometimes spaces. At line 83 I started with spaces. I'll make sure to always use tabs.

erwango replied:
@avisconti , cf #10416, moving all stm32 fixup files to use spaces.

avisconti replied:
I see... OK, I'll use always spaces.

erwango commented on drivers/i2s/Kconfig.stm32:
```diff
@@ -0,0 +1,143 @@
+
+if I2S_STM32
+
+config I2S_STM32_RX_BLOCK_COUNT
```
Is that something that can be tuned?
If yes, can it be put in dts?

avisconti replied:
The DTS should be only used to describe the h/w not the s/w configuration. So, I don't think it is correct to put those flags into Device Tree.

Are they necessary? Well, I'm not sure in fact. I think I can fix them to 4 in the .h. But why not leave the possibility to configure them?

erwango replied:
ok, then keep them that way.

erwango commented on drivers/i2s/Kconfig.stm32:
```diff
@@ -0,0 +1,143 @@
+
+if I2S_1
+
+config I2S_1_NAME
```
Get this from DTS

avisconti replied:
ok, done!

erwango commented on drivers/i2s/Kconfig.stm32:
```diff
@@ -0,0 +1,143 @@
+
+config I2S_1_NAME
+	string "I2S port 1 device name"
+	depends on !HAS_DTS_I2S
```
!HAS_DTS_I2S should not be possible, you can clean matching code

erwango replied:
yes, please.

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1152 @@
+
+static void i2s_stm32_irq_config_func_1(struct device *dev)
+{
+	SYS_LOG_DBG("Enable I2S1 IRQ");
```
Can you clean up this dgb message?
Not sure it adds a real value.

avisconti replied:
ack

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1152 @@
+
+static void i2s_stm32_irq_config_func_2(struct device *dev)
+{
+	SYS_LOG_DBG("Enable I2S2 IRQ");
```
Same

avisconti replied:
ack

erwango commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -0,0 +1,1152 @@
+
+static void i2s_stm32_irq_config_func_3(struct device *dev)
+{
+	SYS_LOG_DBG("Enable I2S3 IRQ");
```
same

erwango reviewed (changes requested):
Some nitpicks, otherwise seems quite nice

erwango reviewed (commented):
One comment/question:
If I2S instances are mapped on SPI ones, how do you ensure both are not active at same time to avoid conflicts?

avisconti commented:
> One comment/question:
> If I2S instances are mapped on SPI ones, how do you ensure both are not active at same time to 
> avoid conflicts?

Good point, but I don't think it has an obvious solution. Maybe we could thing of adding dependencies in Kconfig file for bot I2S and SPI to ensure they are not on at the same time. But not sure it is worth the hassle. When configuring the system one should know that SPI5 must be used as I2S5 and viceversa.

erwango commented on dts/Kconfig:
```diff
@@ -32,6 +32,13 @@ config HAS_DTS_SPI
 	  This option specifies that the target platform supports device tree
 	  configuration for SPI.
 
+config HAS_DTS_I2S
```
Is that used finally?

avisconti replied:
No, I'm removing it.
Give me 10 minutes and I will resubmit the patch

avisconti replied:
@erwango 
done! I addressed all the points

erwango commented on dts/arm/st/f4/stm32f411.dtsi:
```diff
@@ -27,5 +27,25 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40013400 0x400>;
+			interrupts = <84 5>;
```
Miss "clocks"

avisconti replied:
I will add it. But in a separate commit we should probably fix also the spi (clocks is missing also there).
BTW, I wonder how the clocks are currently provided to driver. It seems that currently this 'clocks' declaration in dts is not used yet.

avisconti replied:
Ooooh! OK, got it.
I should add clock there, which means the address + bit at which the clock gating is placed.
I can do that, but then I would also need to fix the driver, as currently I'm configuring the clock gating there, but instead I will obtain it from DTS.

For example now I have in I2S driver:
   806 >       .pclken = {
   807 >       >       .enr = LL_APB2_GRP1_PERIPH_SPI1,
   808 >       >       .bus = STM32_CLOCK_BUS_APB2,
   809 >       },

This part I guess will disappear, as now it will be in DTS.

avisconti replied:
I'll do some tests!

erwango replied:
Yes, indeed, you'll get these from dts as well

avisconti replied:
done for all!

erwango replied:
Well... they should. Did you had troubles when using thoses?

erwango replied:
By the way, you should use the CONFIG_ equivalenst coming from fixup files, not the ST_STM32_...

erwango replied:
So, @avisconti ,  I just created https://github.com/zephyrproject-rtos/zephyr/issues/10435 to get this done one I2C and will do the same for other drivers that don't do it yet.
Would be nice you can do it right now on I2S.

avisconti replied:
@erwango 
OK, agree.  I'm taking a look at it right now.

avisconti replied:
> By the way, you should use the CONFIG_ equivalenst coming from fixup files, 
> not the ST_STM32_...

Yes, yes, of course. I already investigated it yesterday. But I stopped it because it was not done for any IP (I2c, SPI, ...).

erwango commented on dts/arm/st/f4/stm32f411.dtsi:
```diff
@@ -27,5 +27,25 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40015000 0x400>;
+			interrupts = <85 5>;
```
Miss "clocks"

erwango commented on dts/arm/st/f4/stm32f429.dtsi:
```diff
@@ -58,7 +58,17 @@
+			compatible = "st,stm32-i2s";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40013400 0x400>;
```
miss "clocks"

erwango commented on dts/arm/st/f4/stm32f429.dtsi:
```diff
@@ -71,7 +81,17 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40015000 0x400>;
+			interrupts = <85 5>;
```
Miss "clocks"

erwango commented on dts/arm/st/f4/stm32f429.dtsi:
```diff
@@ -83,5 +103,15 @@
+			compatible = "st,stm32-i2s";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40015400 0x400>;
```
miss "clocks"

erwango reviewed (changes requested):
Late catches otherwise sounds good!

avisconti commented:
@erwango 

OK, not everything should be done in the properly way:

1. The DTS description for I2S contains the 'clocks' declaration
2. The dts.fixup file is declaring two new stuff: CLOCK_BITS and CLOCK_BUS, taking the value
     from the defines generated by DTS.
3. The I2S driver makes use of those defines instead of the LL ones

Pls note that I decided to invert the order of two commits, because it makes much more logical
sense. Now I FIRST add I2S into dts, and THEN I add declaration into dts.fixup (otherwise they are 
not existing yet).

erwango commented:
@avisconti 

> OK, not everything should be done in the properly way:

I guess **now** everything is done in the proper way :-D
Seems a last rebase is required and then we're good to go!

avisconti commented:
@erwango 
Yes, but while rebasing I noticed that dma_stm32 driver has been broken by #07ff2d58.
I have solved it easily (was just a typo) but I wonder whether I can add it to this PR. If I won't I guess that
check will not complete right (it will not compile).
Should I add the fix to dma over here?

erwango commented:
@avisconti , let's push a new PR fixing this issue: https://github.com/zephyrproject-rtos/zephyr/issues/10453.
You'll rebase on that. This is painful but the cleanest way

avisconti commented:
Rebased on top of #10454

carlescufi commented:
@avisconti #10454 is merged now

avisconti commented:
recheck

avisconti commented:
@carlescufi 
ready

galak reviewed (commented):
Why do we have #{address,size}-cells for i2s nodes?

avisconti commented:
> Why do we have #{address,size}-cells for i2s nodes?

@galak 
To be honest I copied paiste from the SPI declaration, since the controller is same and I was expecting similarities also here. What are address-cells size-cells used for in dts?

galak commented:
> @galak
> To be honest I copied paiste from the SPI declaration, since the controller is same and I was expecting similarities also here. What are address-cells size-cells used for in dts?

this is used where there are child nodes, and how we address them on a bus.  It doesn't seem like that will be the case w/I2S, but I'm not sure as I'm not that familiar with I2S.

avisconti commented:
> this is used where there are child nodes, and how we address them on a bus. It doesn't seem 
> like that will be the case w/I2S, but I'm not sure as I'm not that familiar with I2S.

OK, got it thx.
I think that I2S as a master may have multiple childs attached to it, for example 1 codec + 1 ADC.
I tried to google for it and I found following example in linux:

arch/arm/boot/dts/rk322x.dtsi:
 146 >   i2s1: i2s1@100b0000 {
 147 >   >   compatible = "rockchip,rk3228-i2s", "rockchip,rk3066-i2s";
 148 >   >   reg = <0x100b0000 0x4000>;
 149 >   >   interrupts = <GIC_SPI 27 IRQ_TYPE_LEVEL_HIGH>;
 150 >   >   #address-cells = <1>;
 151 >   >   #size-cells = <0>;
 152 >   >   clock-names = "i2s_clk", "i2s_hclk";
 153 >   >   clocks = <&cru SCLK_I2S1>, <&cru HCLK_I2S1_8CH>;
 154 >   >   dmas = <&pdma 14>, <&pdma 15>;
 155 >   >   dma-names = "tx", "rx";
 156 >   >   pinctrl-names = "default";
 157 >   >   pinctrl-0 = <&i2s1_bus>;
 158 >   >   status = "disabled";
 159 >   };

So, I would leave them in.

galak commented:
> OK, got it thx.
> I think that I2S as a master may have multiple childs attached to it, for example 1 codec + 1 ADC.
> I tried to google for it and I found following example in linux:
> 
> arch/arm/boot/dts/rk322x.dtsi:
> 146 > i2s1: i2s1@100b0000 {
> 147 > > compatible = "rockchip,rk3228-i2s", "rockchip,rk3066-i2s";
> 148 > > reg = <0x100b0000 0x4000>;
> 149 > > interrupts = <GIC_SPI 27 IRQ_TYPE_LEVEL_HIGH>;
> 150 > > #address-cells = <1>;
> 151 > > #size-cells = <0>;
> 152 > > clock-names = "i2s_clk", "i2s_hclk";
> 153 > > clocks = <&cru SCLK_I2S1>, <&cru HCLK_I2S1_8CH>;
> 154 > > dmas = <&pdma 14>, <&pdma 15>;
> 155 > > dma-names = "tx", "rx";
> 156 > > pinctrl-names = "default";
> 157 > > pinctrl-0 = <&i2s1_bus>;
> 158 > > status = "disabled";
> 159 > };
> 
> So, I would leave them in.

can you see if there's an actual example with child nodes, or at least references to it in DTS binding docs?

avisconti commented:
@galak 
> can you see if there's an actual example with child nodes, or at least references to it in DTS binding docs?

First of all sorry, but I'm not so fluent in device tree stuff. I have always modified what already existing,
but never got into it deeply.

The closest thing that I was able to find in linux is following file:
Documentation/devicetree/bindings/sound/st,stm32-i2s.txt

In that doc there is an example of a i2s device declaration with a single child (i2s2_port):

  30 Example:
  31 sound_card {
  32 >   compatible = "audio-graph-card";
  33 >   dais = <&i2s2_port>;
  34 };
  35
  36 i2s2: audio-controller@40003800 {
  37 >   compatible = "st,stm32h7-i2s";
  38 >   reg = <0x40003800 0x400>;
  39 >   interrupts = <36>;
  40 >   clocks = <&rcc PCLK1>, <&rcc SPI2_CK>, <&rcc PLL1_Q>, <&rcc PLL2_P>;
  41 >   clock-names = "pclk", "i2sclk",  "x8k", "x11k";
  42 >   dmas = <&dmamux2 2 39 0x400 0x1>,
  43            <&dmamux2 3 40 0x400 0x1>;
  44 >   dma-names = "rx", "tx";
  45 >   pinctrl-names = "default";
  46 >   pinctrl-0 = <&pinctrl_i2s2>;
  47
  48 >   i2s2_port: port@0 {
  49 >   >   cpu_endpoint: endpoint {
  50 >   >   >   remote-endpoint = <&codec_endpoint>;
  51 >   >   >   format = "i2s";
  52 >   >   };
  53 >   };
  54 };
  55
  56 audio-codec {
  57 >   codec_port: port@0 {
  58 >   >   codec_endpoint: endpoint {
  59 >   >   >   remote-endpoint = <&cpu_endpoint>;
  60 >   >   };
  61 >   };
  62 };

Moreover, the same document refers to a more general document about device graph:
Documentation/devicetree/bindings/graph.txt

In this graph document it is stated that addess-cells and size-cells are mandatory if multiple ports or endpoints are present:

 113 Required properties
 114 -------------------
 115
 116 If there is more than one 'port' or more than one 'endpoint' node or 'reg'
 117 property present in the port and/or endpoint nodes then the following
 118 properties are required in a relevant parent node:
 119
 120  - #address-cells : number of cells required to define port/endpoint
 121                     identifier, should be 1.
 122  - #size-cells    : should be zero.

For i2s I think that there might be be only one port but multiple endpoints attached to it. But I'm not so sure.

What I can do is to remove addess-cells and size-cells for now. There are multiple things that should be
done in zephyr dts, like describing dmas and clocktree for example. So we may grow it later.

What do you think?

avisconti commented:
@galak 
As discussed with @erwango I removed the addess-cells and size-cells declaration.
Let's start simple. We may add things later.

galak commented:
> As discussed with @erwango I removed the addess-cells and size-cells declaration.
> Let's start simple. We may add things later.

agreed, was going to suggest the same thing
